## 导入数据

In [1]:
import pandas as pd

In [2]:
original_data = pd.read_csv("D://pythontest//实战//化妆品交易.csv")
new_data = original_data.set_index('订单编码',drop=False, append=False, inplace=False, verify_integrity=False)

In [3]:
new_data.head()

订单编码       订单日期    客户编码 所在区域 所在省份      所在地市  商品编号  订购数量 订购单价  \
订单编码                                                                     
D31313  D31313  2019/5/16  S22796   东区  浙江省       台州市  X091   892  214   
D21329  D21329  2019/5/14  S11460   东区  安徽省       宿州市  X005   276  185   
D22372  D22372  2019/8/26  S11101   北区  山西省       忻州市  X078  1450  116   
D31078  D31078   2019/4/8  S10902   北区  吉林省  延边朝鲜族自治州  X025  1834  102   
D32470  D32470  2019/4/11  S18696   北区  北京市       北京市  X010   887   58   

              金额  
订单编码              
D31313  190888.0  
D21329   51060.0  
D22372  168200.0  
D31078  187068.0  
D32470   51446.0

In [4]:
new_data.tail()

订单编码       订单日期    客户编码 所在区域 所在省份 所在地市  商品编号  订购数量 订购单价        金额
订单编码                                                                       
D23184  D23184   2019/8/8  S16891   南区  福建省  泉州市  X003  1493  168  250824.0
D29236  D29236   2019/8/5  S12103   东区  江苏省  扬州市  X059  1592   56   89152.0
D24284  D24284  2019/6/11  S15535   东区  江苏省  南京市  X062   218  239   52102.0
D23538  D23538  2019/2/23  S15969   西区  陕西省  西安市  X031   112   86    9632.0
D21661  D21661   2019/6/9  S13020   南区  湖南省  岳阳市  X100   877   80   70160.0

## 评估数据

这一部分将对在上一部分建立的original_data这个DataFrame所包含的数据进行评估。

评估主要从两个方面进行：结构和内容，即整齐度和干净度。数据的结构性问题指不符合"每列是一个变量,每行是一个观察值，每个单元格是一个值"这三个标准，数据的内容性问题包括存在丢失数据、重复数据、无效数据等。

### 评估数据结构

In [5]:
new_data.head(10)

订单编码       订单日期    客户编码 所在区域     所在省份        所在地市  商品编号  订购数量 订购单价  \
订单编码                                                                           
D31313  D31313  2019/5/16  S22796   东区      浙江省         台州市  X091   892  214   
D21329  D21329  2019/5/14  S11460   东区      安徽省         宿州市  X005   276  185   
D22372  D22372  2019/8/26  S11101   北区      山西省         忻州市  X078  1450  116   
D31078  D31078   2019/4/8  S10902   北区      吉林省    延边朝鲜族自治州  X025  1834  102   
D32470  D32470  2019/4/11  S18696   北区      北京市         北京市  X010   887   58   
D28336  D28336   2019/2/8  S17681   西区      云南省  红河哈尼族彝族自治州  X039  1102  108   
D29820  D29820  2019#3#11  S13780   南区      广东省         江门市  X011  1180  226   
D31449  D31449   2019/5/1  S12455   南区  广西壮族自治区         百色市  X051  1236  203   
D22542  D22542  2019/9/18  S10865   北区   内蒙古自治区       呼和浩特市  X089  1101  104   
D31859  D31859  2019/1/17  S17785   西区      云南省         曲靖市  X001   235  121   

              金额  
订单编码              
D31313  190888.0  
D21329   51060.0  
D22372  168200.0  
D31078  187068.0  
D32470   51446.0  
D28336  119016.0  
D29820  266680.0  
D31449  250908.0  
D22542  114504.0  
D31859   28435.0

从抽样的10行数据来看，数据符合“每列是一个变量,每行是一个观察值，每个单元格是一个值”的标准，每一行是某客户关于某商品的一次交易

### 评估数据干净度

In [6]:
new_data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 31452 entries, D31313 to D21661
Data columns (total 10 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   订单编码    31452 non-null  object 
 1   订单日期    31452 non-null  object 
 2   客户编码    31452 non-null  object 
 3   所在区域    31450 non-null  object 
 4   所在省份    31450 non-null  object 
 5   所在地市    31452 non-null  object 
 6   商品编号    31451 non-null  object 
 7   订购数量    31450 non-null  object 
 8   订购单价    31448 non-null  object 
 9   金额      31448 non-null  float64
dtypes: float64(1), object(9)
memory usage: 2.6+ MB


从输出结果来看，数据共有31452条观测值，所在区域、所在省份、商品编号、订购数量、订购单价、金额存在1-4条缺失值。

此外，订单日期的数据类型应为日期，订购数量和订购单价应为整数或浮点数。

#### 评估缺失数据

查看缺失值，能补充的补一下

In [7]:
new_data[new_data["所在区域"].isnull()]

订单编码       订单日期    客户编码 所在区域 所在省份 所在地市  商品编号  订购数量 订购单价        金额
订单编码                                                                       
D25844  D25844  2019/2/27  S14821  NaN  江苏省  扬州市  X045   538  205  110290.0
D26806  D26806  2019/2/23  S16365  NaN  广东省  湛江市  X022  1542  158  243636.0

In [8]:
new_data.loc["D25844", "所在区域"] = "东区"
new_data.loc["D26806", "所在区域"] = "南区"

In [9]:
new_data[new_data["所在省份"].isnull()]

订单编码       订单日期    客户编码 所在区域 所在省份 所在地市  商品编号  订购数量 订购单价        金额
订单编码                                                                       
D22734  D22734  2019/2/23  S12453   南区  NaN  广州市  X103  1067  202  215534.0
D26028  D26028  2019/7/13  S15878   北区  NaN  北京市  X022  1317  158  208086.0

In [10]:
new_data.loc["D22734", "所在省份"] = "广东省"
new_data.loc["D26028", "所在省份"] = "北京市"

In [11]:
new_data[new_data["订购数量"].isnull()]

订单编码       订单日期    客户编码 所在区域 所在省份 所在地市  商品编号 订购数量 订购单价  金额
订单编码                                                                
D38486  D38486  2019/9/15  S12592   东区  浙江省  杭州市   NaN  NaN  NaN NaN
D31856  D31856  2019/7/12  S17476   东区  湖北省  武汉市  X019  NaN  NaN NaN

In [12]:
new_data[new_data["订购单价"].isnull()]

订单编码       订单日期    客户编码 所在区域 所在省份 所在地市  商品编号  订购数量 订购单价  金额
订单编码                                                                 
D22086  D22086  2019/9/11  S22887   南区  江西省  九江市  X011  1698  NaN NaN
D23111  D23111  2019/4/22  S10909   东区  浙江省  温州市  X096   804  NaN NaN
D38486  D38486  2019/9/15  S12592   东区  浙江省  杭州市   NaN   NaN  NaN NaN
D31856  D31856  2019/7/12  S17476   东区  湖北省  武汉市  X019   NaN  NaN NaN

In [13]:
new_data.dropna(subset= "订购单价", inplace=True)

从输出结果来看，一共有4行无效数据。

#### 评估重复数据

In [14]:
new_data[new_data.duplicated()]

订单编码       订单日期    客户编码 所在区域 所在省份 所在地市  商品编号  订购数量 订购单价        金额
订单编码                                                                       
D33486  D33486  2019/6/30  S15608   南区  广东省  江门市  X049   265  226   59890.0
D23184  D23184   2019/8/8  S16891   南区  福建省  泉州市  X003  1493  168  250824.0
D29236  D29236   2019/8/5  S12103   东区  江苏省  扬州市  X059  1592   56   89152.0
D24284  D24284  2019/6/11  S15535   东区  江苏省  南京市  X062   218  239   52102.0
D23538  D23538  2019/2/23  S15969   西区  陕西省  西安市  X031   112   86    9632.0
D21661  D21661   2019/6/9  S13020   南区  湖南省  岳阳市  X100   877   80   70160.0

In [15]:
new_data.drop_duplicates(keep='first',inplace=True)
new_data

订单编码       订单日期    客户编码 所在区域 所在省份      所在地市  商品编号  订购数量 订购单价  \
订单编码                                                                     
D31313  D31313  2019/5/16  S22796   东区  浙江省       台州市  X091   892  214   
D21329  D21329  2019/5/14  S11460   东区  安徽省       宿州市  X005   276  185   
D22372  D22372  2019/8/26  S11101   北区  山西省       忻州市  X078  1450  116   
D31078  D31078   2019/4/8  S10902   北区  吉林省  延边朝鲜族自治州  X025  1834  102   
D32470  D32470  2019/4/11  S18696   北区  北京市       北京市  X010   887   58   
...        ...        ...     ...  ...  ...       ...   ...   ...  ...   
D22058  D22058  2019/4/16  S10819   西区  四川省       乐山市  X079  1675  128   
D33040  D33040   2019/5/3  S20808   西区  陕西省       宝鸡市  X015   388  151   
D26211  D26211   2019#3#3  S14926   北区  辽宁省       大连市  X014  1166   86   
D23434  D23434  2019/9/27  S18445   北区  河北省       沧州市  X011   976  226   
D27724  D27724   2019/8/6  S15803   西区  贵州省      黔东南州  X038   665  102   

              金额  
订单编码              
D31313  190888.0  
D21329   51060.0  
D22372  168200.0  
D31078  187068.0  
D32470   51446.0  
...          ...  
D22058  214400.0  
D33040   58588.0  
D26211  100276.0  
D23434  220576.0  
D27724   67830.0  

[31442 rows x 10 columns]

In [16]:
new_data[new_data.duplicated()]

Empty DataFrame
Columns: [订单编码, 订单日期, 客户编码, 所在区域, 所在省份, 所在地市, 商品编号, 订购数量, 订购单价, 金额]
Index: []

存在部分重复的观测值，将其删除。

#### 评估不一致数据

不一致数据可能存在于所在省份、所在地市

In [17]:
new_data["所在省份"].value_counts()

江苏省         4731
广东省         3347
浙江省         2645
福建省         1759
四川省         1525
安徽省         1519
湖南省         1483
山东省         1424
上海市         1276
江西省         1077
湖北省         1060
河北省         1047
广西壮族自治区     1041
重庆市          894
云南省          801
陕西省          725
河南省          640
北京市          624
辽宁省          616
山西省          575
贵州省          553
黑龙江省         543
吉林省          380
甘肃省          334
新疆维吾尔自治区     227
天津市          186
海南省          154
内蒙古自治区       140
宁夏回族自治区      116
Name: 所在省份, dtype: int64

In [18]:
new_data["所在地市"].value_counts()

上海市    1276
苏州市    1136
重庆市     894
深圳市     879
泉州市     828
       ... 
白银市       8
周口市       7
邢台市       6
玉溪市       6
崇左市       6
Name: 所在地市, Length: 249, dtype: int64

从输出结果来看，不存在不一致数据。

#### 评估无效或错误数据

In [19]:
new_data.describe()
# new_data["订购数量"]=new_data["订购数量"].astype('float')
# new_data["订购单价"]=new_data["订购单价"].astype('float')
# 发现订购数量里出现了“个”的文字

金额
count   31442.000000
mean   141663.659468
std     99628.479108
min       392.000000
25%     64974.000000
50%    120790.000000
75%    198964.750000
max    779285.000000

In [20]:
new_data[["订购数量1", "订购数量2"]] = new_data["订购数量"].str.split("个", expand = True)
new_data[["订购单价1", "订购单价2"]] = new_data["订购单价"].str.split("元", expand = True)

In [21]:
new_data[new_data["订购数量"]== "1473个"]

订单编码       订单日期    客户编码 所在区域 所在省份 所在地市  商品编号   订购数量 订购单价        金额  \
订单编码                                                                           
D22603  D22603  2019/1/20  S21335   南区  湖南省  衡阳市  X031  1473个   86  126678.0   

       订购数量1 订购数量2 订购单价1 订购单价2  
订单编码                            
D22603  1473          86  None

In [22]:
new_data[new_data["订购单价"]== "86元"]

订单编码       订单日期    客户编码 所在区域 所在省份  所在地市  商品编号 订购数量 订购单价       金额  \
订单编码                                                                         
D25538  D25538  2019/1/21  S14506   北区  河北省  秦皇岛市  X014  397  86元  34142.0   

       订购数量1 订购数量2 订购单价1 订购单价2  
订单编码                            
D25538   397  None    86

In [23]:
new_data["订购数量1"]=new_data["订购数量1"].astype('float')

In [24]:
new_data["订购单价1"]=new_data["订购单价1"].astype('float')

In [25]:
new_data.drop(["订购数量", "订购单价", "订购数量2", "订购单价2"], axis = 1, inplace=True)

In [26]:
new_data = new_data.rename(columns={"订购数量1":"订购数量", "订购单价1": "订购单价"})
new_data

订单编码       订单日期    客户编码 所在区域 所在省份      所在地市  商品编号        金额    订购数量  \
订单编码                                                                            
D31313  D31313  2019/5/16  S22796   东区  浙江省       台州市  X091  190888.0   892.0   
D21329  D21329  2019/5/14  S11460   东区  安徽省       宿州市  X005   51060.0   276.0   
D22372  D22372  2019/8/26  S11101   北区  山西省       忻州市  X078  168200.0  1450.0   
D31078  D31078   2019/4/8  S10902   北区  吉林省  延边朝鲜族自治州  X025  187068.0  1834.0   
D32470  D32470  2019/4/11  S18696   北区  北京市       北京市  X010   51446.0   887.0   
...        ...        ...     ...  ...  ...       ...   ...       ...     ...   
D22058  D22058  2019/4/16  S10819   西区  四川省       乐山市  X079  214400.0  1675.0   
D33040  D33040   2019/5/3  S20808   西区  陕西省       宝鸡市  X015   58588.0   388.0   
D26211  D26211   2019#3#3  S14926   北区  辽宁省       大连市  X014  100276.0  1166.0   
D23434  D23434  2019/9/27  S18445   北区  河北省       沧州市  X011  220576.0   976.0   
D27724  D27724   2019/8/6  S15803   西区  贵州省      黔东南州  X038   67830.0   665.0   

         订购单价  
订单编码           
D31313  214.0  
D21329  185.0  
D22372  116.0  
D31078  102.0  
D32470   58.0  
...       ...  
D22058  128.0  
D33040  151.0  
D26211   86.0  
D23434  226.0  
D27724  102.0  

[31442 rows x 10 columns]

In [27]:
new_data.describe()

金额          订购数量          订购单价
count   31442.000000  31442.000000  31442.000000
mean   141663.659468    885.048502    160.285732
std     99628.479108    495.953226     59.976461
min       392.000000      7.000000     56.000000
25%     64974.000000    504.000000    102.000000
50%    120790.000000    859.000000    159.000000
75%    198964.750000   1219.000000    213.000000
max    779285.000000   3210.000000    253.000000

In [28]:
new_data["总金额"] = new_data["订购数量"] * new_data["订购单价"]

In [29]:
new_data[new_data["总金额"]!=new_data["金额"]]
# 不存在数据计算错误的观测值

Empty DataFrame
Columns: [订单编码, 订单日期, 客户编码, 所在区域, 所在省份, 所在地市, 商品编号, 金额, 订购数量, 订购单价, 总金额]
Index: []

不存在无效数据，但出现了不一致数据。

In [30]:
new_data.drop(["金额"], axis = 1, inplace=True)

## 清理数据

目前剩下的步骤为：将订单日期转变为日期类型

In [31]:
cleaned_data = new_data.copy()

In [32]:
# cleaned_data["订单日期"] = pd.to_datetime(cleaned_data["订单日期"])
# 发现有的斜杠变成了#

In [33]:
cleaned_data

订单编码       订单日期    客户编码 所在区域 所在省份      所在地市  商品编号    订购数量   订购单价  \
订单编码                                                                         
D31313  D31313  2019/5/16  S22796   东区  浙江省       台州市  X091   892.0  214.0   
D21329  D21329  2019/5/14  S11460   东区  安徽省       宿州市  X005   276.0  185.0   
D22372  D22372  2019/8/26  S11101   北区  山西省       忻州市  X078  1450.0  116.0   
D31078  D31078   2019/4/8  S10902   北区  吉林省  延边朝鲜族自治州  X025  1834.0  102.0   
D32470  D32470  2019/4/11  S18696   北区  北京市       北京市  X010   887.0   58.0   
...        ...        ...     ...  ...  ...       ...   ...     ...    ...   
D22058  D22058  2019/4/16  S10819   西区  四川省       乐山市  X079  1675.0  128.0   
D33040  D33040   2019/5/3  S20808   西区  陕西省       宝鸡市  X015   388.0  151.0   
D26211  D26211   2019#3#3  S14926   北区  辽宁省       大连市  X014  1166.0   86.0   
D23434  D23434  2019/9/27  S18445   北区  河北省       沧州市  X011   976.0  226.0   
D27724  D27724   2019/8/6  S15803   西区  贵州省      黔东南州  X038   665.0  102.0   

             总金额  
订单编码              
D31313  190888.0  
D21329   51060.0  
D22372  168200.0  
D31078  187068.0  
D32470   51446.0  
...          ...  
D22058  214400.0  
D33040   58588.0  
D26211  100276.0  
D23434  220576.0  
D27724   67830.0  

[31442 rows x 10 columns]

In [34]:
# 拆分
#cleaned_data[["year", "month", "day"]] = cleaned_data["订单日期"].str.split("#", expand = True)

In [35]:
# 拼接
#cleaned_data["year"] = cleaned_data["year"].str.cat(cleaned_data[["month", "day"]], sep = "/")

In [36]:
# for var in cleaned_data["订单编码"]:
#     if cleaned_data.loc['var', "month"]!=None:
#         cleaned_data["date"] = cleaned_data["year"]
#     else:
#         cleaned_data["date"] = cleaned_data["订单日期"]

In [37]:
# cleaned_data.loc["D31313", "month"]

In [38]:
#translation_table = cleaned_data["订单日期"].maketrans("#", "/")

In [39]:
def rep(text):
    return text.replace('#', '/')

In [40]:
date = cleaned_data.loc[:, "订单日期"].apply(rep)
date

订单编码
D31313    2019/5/16
D21329    2019/5/14
D22372    2019/8/26
D31078     2019/4/8
D32470    2019/4/11
            ...    
D22058    2019/4/16
D33040     2019/5/3
D26211     2019/3/3
D23434    2019/9/27
D27724     2019/8/6
Name: 订单日期, Length: 31442, dtype: object

In [41]:
cleaned_data["订单日期"] = date

In [42]:
cleaned_data

订单编码       订单日期    客户编码 所在区域 所在省份      所在地市  商品编号    订购数量   订购单价  \
订单编码                                                                         
D31313  D31313  2019/5/16  S22796   东区  浙江省       台州市  X091   892.0  214.0   
D21329  D21329  2019/5/14  S11460   东区  安徽省       宿州市  X005   276.0  185.0   
D22372  D22372  2019/8/26  S11101   北区  山西省       忻州市  X078  1450.0  116.0   
D31078  D31078   2019/4/8  S10902   北区  吉林省  延边朝鲜族自治州  X025  1834.0  102.0   
D32470  D32470  2019/4/11  S18696   北区  北京市       北京市  X010   887.0   58.0   
...        ...        ...     ...  ...  ...       ...   ...     ...    ...   
D22058  D22058  2019/4/16  S10819   西区  四川省       乐山市  X079  1675.0  128.0   
D33040  D33040   2019/5/3  S20808   西区  陕西省       宝鸡市  X015   388.0  151.0   
D26211  D26211   2019/3/3  S14926   北区  辽宁省       大连市  X014  1166.0   86.0   
D23434  D23434  2019/9/27  S18445   北区  河北省       沧州市  X011   976.0  226.0   
D27724  D27724   2019/8/6  S15803   西区  贵州省      黔东南州  X038   665.0  102.0   

             总金额  
订单编码              
D31313  190888.0  
D21329   51060.0  
D22372  168200.0  
D31078  187068.0  
D32470   51446.0  
...          ...  
D22058  214400.0  
D33040   58588.0  
D26211  100276.0  
D23434  220576.0  
D27724   67830.0  

[31442 rows x 10 columns]

In [44]:
# 日期类型字符串！
cleaned_data["订单日期"] = pd.to_datetime(cleaned_data["订单日期"])
cleaned_data["订单日期"]

订单编码
D31313   2019-05-16
D21329   2019-05-14
D22372   2019-08-26
D31078   2019-04-08
D32470   2019-04-11
            ...    
D22058   2019-04-16
D33040   2019-05-03
D26211   2019-03-03
D23434   2019-09-27
D27724   2019-08-06
Name: 订单日期, Length: 31442, dtype: datetime64[ns]

In [ ]:
# 注意：如果将浮点数转换为字符串，可能有.0结尾，这个时候用var.str.slice(0, -2)

## 保存清理后数据

In [45]:
cleaned_data.to_csv("cosmetics_cleaned.csv", index=False)

In [46]:
pd.read_csv("cosmetics_cleaned.csv").head()

订单编码        订单日期    客户编码 所在区域 所在省份      所在地市  商品编号    订购数量   订购单价  \
0  D31313  2019-05-16  S22796   东区  浙江省       台州市  X091   892.0  214.0   
1  D21329  2019-05-14  S11460   东区  安徽省       宿州市  X005   276.0  185.0   
2  D22372  2019-08-26  S11101   北区  山西省       忻州市  X078  1450.0  116.0   
3  D31078  2019-04-08  S10902   北区  吉林省  延边朝鲜族自治州  X025  1834.0  102.0   
4  D32470  2019-04-11  S18696   北区  北京市       北京市  X010   887.0   58.0   

        总金额  
0  190888.0  
1   51060.0  
2  168200.0  
3  187068.0  
4   51446.0